### PR TITLE
[Feature] Support NPUGraph for DeepSeek on Ascend NPU

### PR DESCRIPTION
**Pull request by chenxu140**
## Motivation

this pr improves deepseek model (mla to be exact) performance with npugraph support
check initial npugraph support on mha model here #9399 and #8030

## Modifications

- support concurrent kvcache transfer following mooncake transfer engine
- add mla graph support in ascend attention backend
- some performance and accuracy improvements

## Accuracy Tests

87DF5544-5E77-4E8C-F895-71FDFDFEEEEB

## Benchmarking and Profiling

Reusing performance result [here](https://github.com/sgl-project/sglang/pull/8328#issuecomment-3222527169)
FIA result: sglang/test/srt/ascend/test_ascend_mla_fia_w8a8int8.py (tp == 2)
image

After this pr, with cuda graph enabled
```bash
python3 -m sglang.launch_server --model DeepSeek-V2-Lite-W8A8 --tp 2 --trust-remote-code --quantization w8a8_int8  --device npu --attention-backend ascend --cuda-graph-bs 1 2 4 6 8 16 24 32 40 48 56 64 72 80 88 96 104 112 120 128 136 --disable-radix-cache --chunked-prefill-size 32768 --mem-fraction-static 0.8
```
D0A19C18-A5A9-4ED0-B5B6-8B5F1BCB3246

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/managers/data_parallel_controller.py [hunk 1]
```diff
@@ -281,7 +284,21 @@ def round_robin_scheduler(self, req: Req):
                 logger.debug(f"Direct routing to DP rank {req.data_parallel_rank}")
                 self.workers[req.data_parallel_rank].send_pyobj(req)
             else:
-                self.workers[req.bootstrap_room % len(self.workers)].send_pyobj(req)
+                DP_LOAD_BALANCE = os.getenv("DP_LOAD_BALANCE", "0") == "1"
```
Comment (iforgetmyname) [inline python/sglang/srt/managers/data_parallel_controller.py]: Revert this block

File: python/sglang/srt/managers/data_parallel_controller.py [hunk 2]
```diff
@@ -101,6 +102,8 @@ def __init__(
         # Launch data parallel workers
         self.scheduler_procs = []
         self.workers = [None] * server_args.dp_size
+        self.work_req_num = [0] * server_args.dp_size
```
Comment (iforgetmyname) [inline python/sglang/srt/managers/data_parallel_controller.py]: why we need to do dp load balance here

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -894,7 +894,7 @@ def __init__(
         start_layer: Optional[int] = None,
         end_layer: Optional[int] = None,
     ):
-        super(MLATokenToKVPool, self).__init__(
+        super(MHATokenToKVPool, self).__init__(
```
Comment (iforgetmyname) [inline python/sglang/srt/mem_cache/memory_pool.py]: use super().__init__ instead as this is a direct inheritance

File: python/sglang/srt/disaggregation/ascend/conn.py [hunk 4]
```diff
@@ -29,6 +35,77 @@ def register_buffer_to_engine(self):
...
+        layers_params = [
+            (
+                self.kv_args.kv_data_ptrs[layer_id],
+                dst_kv_ptrs[layer_id],
+                self.kv_args.kv_item_lens[layer_id],
+            )
+            for layer_id in range(num_layers)
+        ]
+
+        assert layers_params is not None
```
Comment (iforgetmyname) [inline python/sglang/srt/disaggregation/ascend/conn.py]: remove this line

File: python/sglang/srt/managers/data_parallel_controller.py (hunk 1 above)
Comment (chenxu140) [reply python/sglang/srt/managers/data_parallel_controller.py]: done

File: python/sglang/srt/managers/data_parallel_controller.py (hunk 2 above)
Comment (chenxu140) [reply python/sglang/srt/managers/data_parallel_controller.py]: done

File: python/sglang/srt/disaggregation/ascend/conn.py (hunk 4 above)
Comment (chenxu140) [reply python/sglang/srt/disaggregation/ascend/conn.py]: done

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 3 above)
Comment (chenxu140) [reply python/sglang/srt/mem_cache/memory_pool.py]: done

**Comment by iforgetmyname** [comment]
image

this is the comment i deleted here #8328, and it should be fixed in this pr

File: python/sglang/srt/disaggregation/ascend/conn.py [hunk 5]
```diff
@@ -29,6 +35,75 @@ def register_buffer_to_engine(self):
             self.kv_args.aux_data_ptrs, self.kv_args.aux_data_lens
         )
 
+    def send_kvcache(
```
Comment (ErvinXie) [inline python/sglang/srt/disaggregation/ascend/conn.py:38]: Most of the function is identical to mooncacke/conn.py, could we find a way to avoid this?

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 6]
```diff
@@ -54,13 +96,12 @@ def __init__(self, model_runner: ModelRunner):
         super().__init__()
         self.forward_metadata = None
         self.device = model_runner.device
-        self.gen_attention_mask(128, model_runner.dtype)
+        self.attn_mask_builder = AttentionMaskBuilder(8192, model_runner.dtype)
```
Comment (ErvinXie) [inline python/sglang/srt/layers/attention/ascend_backend.py]: can we use compressed mask as described in  https://www.hiascend.com/document/detail/zh/canncommercial/81RC1/apiref/ascendtbapi/ascendtb_01_0259.html

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 7]
```diff
@@ -297,6 +297,7 @@ class ForwardBatch:
     # For padding
     padded_static_len: int = -1  # -1 if not padded
     num_token_non_padded: Optional[torch.Tensor] = None  # scalar tensor
+    num_token_non_padded_cpu: int = None
```
Comment (Alcanderian) [inline python/sglang/srt/model_executor/forward_batch_info.py]: Unused variable

Comment (chenxu140) [reply python/sglang/srt/model_executor/forward_batch_info.py]: fixed

**Comment by Alcanderian** [comment]
CI passed
image

**Review by Alcanderian — APPROVED** [review]

**Comment by Alisehen** [comment]
We profiled the decode time of a single layer, and the performance improvement is significant.
截屏2025-08-28 16 01 22

**Review by ErvinXie — APPROVED** [review]
After thorough testing, we confirmed that enabling NPUGraph delivers a 6x improvement in decode performance. In addition, it shows a clear performance advantage compared to vllm ascend. Based on these results, we believe this PR is ready to be merged.

File: python/sglang/srt/layers/attention/ascend_backend.py (hunk 6 above)
Comment (iforgetmyname) [reply python/sglang/srt/layers/attention/ascend_backend.py]: atb will stop maintaining by the end of this year and also it does't work well with aclnn ops

File: python/sglang/srt/disaggregation/ascend/conn.py (hunk 5 above)
Comment (iforgetmyname) [reply python/sglang/srt/disaggregation/ascend/conn.py:38]: directly inherit from mooncake and add logics to parent method could also be a pain in the butt, we will try to find if there's a way out

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 8]
```diff
@@ -158,7 +158,7 @@ def init_forward_metadata_replay_cuda_graph(
         self.graph_mode = True
 
     def get_cuda_graph_seq_len_fill_value(self):
-        return 1
+        return 0
```
Comment (wangtiance) [inline python/sglang/srt/layers/attention/ascend_backend.py:161]: Hello, anyone care to explain why the value is changed to 0 here? Thanks.
